src/MALA.jl

```diff
@@ -0,0 +1,107 @@
+using .ForwardDiff: gradient!
+using .DiffResults: GradientResult, value, gradient
+
+struct MALA{D} <: MHSampler
+    proposal::D
+end
+
+
+# Create a RandomWalkProposal if we weren't given one already.
+MALA(d) = MALA(RandomWalkProposal(d))
+
+# If we were given a RandomWalkProposal, just use that instead.
+MALA(d::RandomWalkProposal) = MALA{typeof(d)}(d)
+
+
+struct GradientTransition{T<:Union{Vector, Real, NamedTuple}, L<:Real, G<:Union{Vector, Real, NamedTuple}}
+    params :: T
+    lp :: L
+    gradient :: G
+end
+
+
+# Store the new draw, its log density and its gradient
+GradientTransition(model::DensityModel, params) = GradientTransition(params, logdensity_and_gradient(model, params)...)
+
+
+function propose(
+    rng::Random.AbstractRNG,
+    spl::MALA{<:Proposal},
+    model::DensityModel,
+    params_prev::GradientTransition
+)
+    proposal = propose(rng, spl.proposal(params_prev.gradient), model, params_prev.params)
+    return GradientTransition(model, proposal)
+end
+
+
+function q(
+    spl::MALA{<:Proposal},
+    t::GradientTransition,
+    t_cond::GradientTransition
+)
+    return q(spl.proposal(-t_cond.gradient), t.params, t_cond.params)
+end
+
+
+"""
+    logdensity_and_gradient(model::DensityModel, params)
+
+Efficiently returns the value and gradient of the model
+"""
+function logdensity_and_gradient(model::DensityModel, params)
+    res = GradientResult(params)
+    gradient!(res, model.logdensity, params)
+    return (value(res), gradient(res))
+end
+
+
+logdensity(model::DensityModel, t::GradientTransition) = t.lp
+
+
+# Define the first step! function, which is called at the 
+# beginning of sampling. Return the initial parameter used
+# to define the sampler.
+function AbstractMCMC.step!(
+    rng::Random.AbstractRNG,
+    model::DensityModel,
+    spl::MALA{<:Proposal},
+    N::Integer,
+    ::Nothing;
+    init_params=nothing,
+    kwargs...
+)
+    if init_params === nothing
+        @warn "need init_params"
+    else
+        return GradientTransition(model, init_params)
+    end
+end
+
+
+
+# Define the other step functions. Returns a Transition containing
+# either a new proposal (if accepted) or the previous proposal 
+# (if not accepted).
+function AbstractMCMC.step!(
+    rng::Random.AbstractRNG,
+    model::DensityModel,
+    spl::MALA{<:Proposal},
+    ::Integer,
+    params_prev;
+    kwargs...
+)
+    # Generate a new proposal.
+    params = propose(rng, spl, model, params_prev)
+
+    # Calculate the log acceptance probability.
+    logα = logdensity(model, params) - logdensity(model, params_prev) +
+        q(spl, params_prev, params) - q(spl, params, params_prev)
+
+    # Decide whether to return the previous params or the new one.
+    if -Random.randexp(rng) < logα
+        return params
+    else
+        return params_prev
+    end
+end
```

Comment:
I suggest the following:
- Remove these two functions:
```suggestion
```
- Add the lines
```julia
transition(sampler, model, params) = transition(model, params)
transition(model, params) = Transition(model, params)
```
to mh-core.jl
- Replace https://github.com/TuringLang/AdvancedMH.jl/blob/2bad97c703b6ae44fc8af5741e9ee36a01170671/src/mh-core.jl#L185 with
```julia
        transition = AdvancedMH.transition(spl, model, init_params)
```
- Add the following lines below the constructor of `GradientTransition` in MALA.jl:
```julia
transition(::MALA, model, params) = GradientTransition(model, params)
```

Then you should not have to care about AbstractMCMC anymore and could use the default implementation and would get the correct transition.

(We might want to change `Transition` to `AdvancedMH.transition` in more places to unify the implementations but this can be done in subsequent PRs)

Reply:
Thanks for this. 

In addition to these changes, I had to change some of the input types for the `AbstractMCMC.step` methods in `mh-core.jl`.

```julia
     function AbstractMCMC.step(
    rng::Random.AbstractRNG,
    model::DensityModel,
    spl::MHSampler,
    params_prev::AbstractTransition;
    kwargs...
)
```
I had to change the type for `spl` from `MetropolisHastings` to `MHSampler` to allow both `MetropolisHastings` and `MALA` types, and I had to introduce a new abstract class in `AdvancedMH.jl`

```julia
     abstract type AbstractTransition end
```

and specify that `params_prev` in `AbstractMCMC.step` is of that type so that it accepts both `Transition` and `GradientTransition`.

I also had to change the type of `ts` in `AbstractMCMC.bundle_samples` to accept `AbstractTransition`:

```julia
AbstractMCMC.bundle_samples(
    ts::Vector{<:AbstractTransition},
    model::DensityModel,
    sampler::MHSampler,
    state,
    chain_type::Type{Vector{NamedTuple}};
    param_names=missing,
    kwargs...
)
```

Let me know if you think these changes are OK.

My apologies- this is the first time I'm submitted my work to another repo including others work, so I'm just getting used to the whole joint-coding thing, and in general I'm more of a "ham and egger"- the quality of the code here is way above my standard :) but I'm trying....

Reply:
Yes, you are completely right! I missed that the type signature was too restrictive.